lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+	if !currentEpochEntry.ContainsBlockHeight(block.Header.Height) {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("block is not in current epoch"))
+	}
+	snapshotAtEpochNumber, err := utxoView.ComputeSnapshotEpochNumberForEpoch(currentEpochEntry.EpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem getting snapshot epoch number for epoch #%d", currentEpochEntry.EpochNumber)
+	}
+	isValidPartialSig, err := utxoView.hasValidProposerPartialSignaturePoS(block, snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrap(err, "validateAndIndexBlockPoS: Problem validating proposer partial sig")
+	}
+	if !isValidPartialSig {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("invalid proposer partial sig"))
+	}
+	// A couple of options on how to handle:
+	//   - Add utility to UtxoView to fetch the validator set given an arbitrary block height. If we can't fetch the
+	//     validator set for the block, then we reject it (even if it later turns out to be a valid block)
+	//   - Add block to the block index before QC validation such that even if we aren't able to fetch the validator
+	//     set for the block, we can at least store it locally.
+	// 2. Validate QC
+	validatorsByStake, err := utxoView.GetAllSnapshotValidatorSetEntriesByStake()
+	if err != nil {
+		// This should never happen. If the parent is validated and extends from the tip, then we should
+		// be able to fetch the validator set at its block height for it. This failure can only happen due
+		// to transient badger issues. We return the block node as-is as a best effort thing.
+		return blockNode, nil
+	}
+
+	// Validate the block's QC. If it's invalid, we store it as ValidateFailed.
+	if err = bc.isValidPoSQuorumCertificate(block, validatorsByStake); err != nil {
+		return bc.storeValidateFailedBlockWithWrappedError(block, err)
+	}
+
+	isBlockProposerValid, err := utxoView.hasValidBlockProposerPoS(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem validating block proposer")
+	}
+	if !isBlockProposerValid {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("invalid block proposer"))
+	}
+
+	// Connect this block to the parent block's UtxoView.
+	txHashes := collections.Transform(block.Txns, func(txn *MsgDeSoTxn) *BlockHash {
+		return txn.Hash()
+	})
+
+	// If we fail to connect the block, then it means the block is invalid. We should store it as ValidateFailed.
+	if _, err = utxoView.ConnectBlock(block, txHashes, true, nil, block.Header.Height); err != nil {
+		// If it doesn't connect, we want to mark it as ValidateFailed.
+		return bc.storeValidateFailedBlockWithWrappedError(block, err)
+	}
+
+	// We can now add this block to the block index since we have performed all basic validations.
+	blockNode, err = bc.storeValidatedBlockInBlockIndex(block)
```

Comment:
So at this point you store the block and mark it as validated. But now suppose the node shuts down right here and reboots. Wouldn't this result in an inconsistent state?

I think you need to mark the block as validated in the same txn where you update the tip. Otherwise, you can get weird edge cases where a new block comes in and you think you've processed this one but you haven't actually.

Thaat would mean that you should just store it here but not mark it as validated.

Reply:
Hey sorry I realized that this is OK because of the changes we're making with PoS. Namely: We don't update badger with all the connected transactions until we COMMIT, which means that it's correct to mark the block as validated here and store it.